### PR TITLE
maxwell_3d: Implement MME shadow RAM

### PR DESCRIPTION
**Pull request by blueskythlikesclouds**
Implementation was based on this Ryujinx PR: https://github.com/Ryujinx/Ryujinx/pull/987
Makes Hatsune Miku: Project DIVA MEGA39's render correctly.

### CODE REVIEW

File: src/video_core/engines/maxwell_3d.h [hunk 1]
```diff
@@ -674,7 +681,9 @@ class Maxwell3D final : public ConstBufferEngineInterface {
                     u32 bind;
                 } macros;
 
-                INSERT_UNION_PADDING_WORDS(0x17);
+                ShadowRamControl shadow_ram_control;
```
Comment (ReinUsesLisp) [inline src/video_core/engines/maxwell_3d.h]: This is missing the static assertion at the bottom of this file.

Comment (blueskythlikesclouds) [reply src/video_core/engines/maxwell_3d.h]: OK, I added it.

File: src/video_core/engines/maxwell_3d.cpp [hunk 2]
```diff
@@ -121,10 +121,19 @@ void Maxwell3D::CallMacroMethod(u32 method, std::size_t num_parameters, const u3
...
+    u32 argument = method_call.argument;
+
+    if (method >= 0x60 && method < MacroRegistersStart) {
+        if (regs.shadow_ram_control == Regs::ShadowRamControl::Replay) {
+            argument = shadow_regs.reg_array[method];
+        } else if (regs.shadow_ram_control == Regs::ShadowRamControl::Track ||
+                   regs.shadow_ram_control == Regs::ShadowRamControl::TrackWithFilter) {
+            shadow_regs.reg_array[method] = argument;
+        }
+    }
```
Comment (FernandoS27) [inline src/video_core/engines/maxwell_3d.cpp]: This should be checked on the macro interpreter, not on method call.

**Comment by FernandoS27** [comment]
Also creditting should go to fincs who REd the MMEShadow memory.

File: src/video_core/engines/maxwell_3d.cpp (hunk 2 above)
Comment (blueskythlikesclouds) [reply src/video_core/engines/maxwell_3d.cpp]: I suppose in the MacroInterpreter::Send method? Will also probably need to make get/set methods for shadow registers.

Comment (FernandoS27) [reply src/video_core/engines/maxwell_3d.cpp]: Feel free to do so or use class friendliness.

Comment (blueskythlikesclouds) [reply src/video_core/engines/maxwell_3d.cpp]: I've kept the handler for Track and TrackWithFilter in Maxwell3D::CallMethod but moved the handler for Replay to MacroInterpreter::Send, is that fine? Moving both to macro interpreter makes Vulkan crash and kills 3D rendering in OpenGL.

Like so:
```cpp
void Maxwell3D::CallMethod(const GPU::MethodCall& method_call) {
    const u32 method = method_call.method;

    // Keep track of the register value in shadow_regs when requested.
    if (method < Regs::NUM_REGS &&
        (regs.shadow_ram_control == Regs::ShadowRamControl::Track ||
         regs.shadow_ram_control == Regs::ShadowRamControl::TrackWithFilter)) {
        shadow_regs.reg_array[method] = method_call.argument;
    }
```

```cpp
void MacroInterpreter::Send(u32 value) {
    // Use the tracked value in shadow_regs when requested.
    if (method_address.address < Engines::Maxwell3D::Regs::NUM_REGS &&
        maxwell3d.regs.shadow_ram_control == Engines::Maxwell3D::Regs::ShadowRamControl::Replay) {
        value = maxwell3d.shadow_regs.reg_array[method_address.address];
    }
```

Not sure if I should move the check to somewhere else in the CallMethod function.

**Comment by slashiee** [comment]
Before:


After:


<details>
<summary>This also fixes missing rendering in Bayonetta.</summary>

Before:

After:

</details>

**Comment by blueskythlikesclouds** [comment]
@FernandoS27 Where should I include the credits for Ryujinx and fincs? Above `enum class ShadowRamControl` or at the beginning of file? (after the copyright notice)

**Comment by bunnei** [comment]
Thanks for making this PR! Users have reported that it is breaking a few games (e.g. SMO is quite broken, and perhaps others). Can you please take a look? Thank you!

**Comment by blueskythlikesclouds** [comment]
Unfortunately, I was unable to fix the problem. Someone who actually knows what they are doing should implement this. I'm closing the PR.

**Pull request closed without merging**